tests/task_scheduler_test.c

```diff
@@ -198,7 +198,63 @@ static int test_scheduler_rejects_xthread_access(struct aws_allocator *alloc, vo
     return 0;
 }
 
+/* container for running the test making sure a recursive call to aws_task_scheduler_schedule_now
+ * does not break the fairness of the task scheduler. */
+struct task_scheduler_reentrancy_args {
+    struct aws_task_scheduler *scheduler;
+    int executed;
+    struct task_scheduler_reentrancy_args *next_task_args;
+};
+
+static void reentrancy_fn(void *arg) {
+
+    struct task_scheduler_reentrancy_args *reentrancy_args = (struct task_scheduler_reentrancy_args *)arg;
+
+    if (reentrancy_args->next_task_args) {
+        struct aws_task task;
+        task.fn = reentrancy_fn;
+        task.arg = reentrancy_args->next_task_args;
+
+        aws_task_scheduler_schedule_now(reentrancy_args->scheduler, &task);
+    }
+
+    reentrancy_args->executed = 1;
+}
+
+static int test_scheduler_reentrant_safe(struct aws_allocator *alloc, void *ctx) {
+    struct aws_task_scheduler scheduler;
+    aws_task_scheduler_init(&scheduler, alloc, aws_high_res_clock_get_ticks);
+
+    struct task_scheduler_reentrancy_args task2_args;
+    task2_args.scheduler = &scheduler;
+    task2_args.executed = 0;
+    task2_args.next_task_args = NULL;
+
+    struct task_scheduler_reentrancy_args task1_args;
+    task1_args.scheduler = &scheduler;
+    task1_args.executed = 0;
+    task1_args.next_task_args = &task2_args;
+
+    struct aws_task task;
+    task.arg = &task1_args;
+    task.fn = reentrancy_fn;
+
+    ASSERT_SUCCESS(aws_task_scheduler_schedule_now(&scheduler, &task), "Failed while scheduling task.");
+
+    ASSERT_SUCCESS(aws_task_scheduler_run_all(&scheduler, NULL), "Executing tasks failed.");
+
+    ASSERT_TRUE(task1_args.executed, "Task 1 should have been executed but was not.");
+    ASSERT_FALSE(task2_args.executed, "Task 2 should not have been executed but was.");
+
+    ASSERT_SUCCESS(aws_task_scheduler_run_all(&scheduler, NULL), "Executing tasks failed.");
+    ASSERT_TRUE(task2_args.executed, "Task 2 should not have been executed but was.");
```

Comment:
Wrong assert string.

As an aside, I'd suggest leaving off assert strings when they state the same thing as the actual assert expression - the default assert message will print e.g. `Expected condition to be true: task2_args.executed [task_scheduler.test.c:250]`, which says the same thing as your assert string would say (if it wasn't wrong due to a copy-paste error), but with less typing and no chance of being inconsistent with the actual condition being tested.

Reply:
I'm going to add a coding style/standards section to the README. 

But the guidance I think I'd like to go with is,

Use the strings when you want to give reasoning for why x is a failure condition, but not just to print out the failure condition.